examples/gbasf2_fei_training/fei_grid_workflow.py

```diff
@@ -0,0 +1,473 @@
+            for key in self.get_input_file_names():
+                if key == "mcParticlesCount.root" or key == "training_input.root" or "Monitor" in key or key.endswith(".xml"):
+                    adjusted_key = '"' + key + '"'
+                    os.system(f"rm {adjusted_key}")
+            os.system("rm -f Summary.pickle*")
+
+            if self.stage < 6:
+                # move *.xml and *.log files to output directory
+                if len(glob.glob("*.xml")) > 0:
+                    os.system(f'mv *.xml {outputdir}')
+                if len(glob.glob("*.log")) > 0:
+                    os.system(f'mv *.log {outputdir}')
+
+
+class PrepareInputsTask(luigi.Task):
+
+    remote_tmp_directory = luigi.Parameter(significant=False)  # should be set via settings
+    remote_initial_se = luigi.Parameter(significant=False)  # should be set via settings
+
+    stage = luigi.IntParameter()
+    mode = luigi.Parameter()
+
+    def output(self):
+
+        yield self.add_to_output('fei_analysis_inputs.tar.gz')
+        yield self.add_to_output('successfull_input_upload.txt')
+
+    def requires(self):
+
+        # need merged mcParticlesCount.root from stage -1
+        yield MergeOutputsTask(
+            mode="Merging",
+            stage=-1,
+            ncpus=luigi.get_setting("local_cpus"),
+        )
+
+        # need dataset_sites.txt file from stage -1 of FEI training,
+        # which is (ab)used to create that site list
+        yield FEITrainingTask(
+            mode="Training",
+            stage=-1,
+        )
+
+        # need .xml training files from current stage of FEI training
+        if self.stage > -1:
+            yield FEITrainingTask(
+                mode="Training",
+                stage=self.stage,
+            )
+
+        # need .xml training files from all previous stages of FEI training,
+        # beginning with stage 1
+        if self.stage > 0:
+
+            for fei_stage in range(self.stage):
+
+                yield FEITrainingTask(
+                    mode="Training",
+                    stage=fei_stage,
+                )
+
+    def run(self):
+
+        # create tarball with all required input files for FEIAnalysisTask
+        outputs = [outs[0] for outs in self.get_input_file_names().values()
+                   if outs[0].endswith('.root') or outs[0].endswith('.xml')]
+        taroutname = self.get_output_file_name("fei_analysis_inputs.tar.gz")
+        taroutdir = os.path.dirname(taroutname)
+        outputsstring = ' '.join(['"'+o+'"' for o in outputs])
+        tarcmd = f"rm -f {self.get_output_file_name('successfull_input_upload.txt')}; "
+        tarcmd += f"cp {outputsstring} {taroutdir}; "
+        tarcmd += f"pushd {taroutdir}; "
+        baseoutputsstring = ' '.join(['"'+os.path.basename(o)+'"' for o in outputs])
+        tarcmd += f"tar -vczf {taroutname} {baseoutputsstring}; "
+        tarcmd += f"rm -f {baseoutputsstring}; popd"
+        os.system(tarcmd)
```

Comment:
This bash programming from python is difficult to read for me, even if it works when you first hack together.  I would replace the calls to shell command with proper python. Why use bash when we have the power of python? Use the built-in [tarfile](https://docs.python.org/3.7/library/tarfile.html) package for unpacking, `shutil.copy` for copying and `os.remove` for removing etc. Using python ensures that the code runs for users with non-bash shells and on other operating systems (though officially we only support linux anyway). Also it ensures proper error handling in case one of the command fails, and the user will see which python function exactly failed. Maybe if I take the time to understand what exactly the bash commands are doing I will replace it myself.

Anyway, since this is just an example and not part of the b2luigi codebase I'm not that strict here. At the very least I would replace `os.system` with the newer `subprocess.call(shell=True)` (I think your commands with `pushd`, `popd` etc. demand a shell, if you don't change them).

Reply:
The aim is a tarball containing the needed input files from different directories **without** their directory structure (just the files), and the tarball should be put into the output directory of the task.

I had a few trials with the `tarfile` package, but I haven't managed to have it done as I needed it to be. The shell command construction was the nearest solution to overcome that.

Reply:
I'll try to rethink this a bit to solve it

Reply:
Actually maybe it's easier to do with [shutil.make_archive](https://docs.python.org/3.7/library/shutil.html#shutil.make_archive):
``` python
        shutil.make_archive(
            base_name=<base name of output tar file without the file ending>,
            format="gztar",
            root_dir=<directory in which the tar.gz archive should be created>,
            base_dir=<directory to archive>,
            verbose=True,
            dry_run=False,
        )
```
I looked at my quality indicator training script which is in basf2 and I had used this function to make an archive (see [here](https://stash.desy.de/projects/b2/repos/software/browse/tracking/scripts/tracking/train/combined_quality_estimator_teacher.py#2452-2458))

Reply:
I actually have now a solution with `tarfile` :D, at least I think it will work out

Reply:
Let me put it together with `tarfile`, and then we can decide what's better :)

Reply:
Ok I think in that way, it should work out: https://github.com/ArturAkh/b2luigi/blob/fei_with_gbasf2/examples/gbasf2_fei_training/fei_grid_workflow.py#L410-L419